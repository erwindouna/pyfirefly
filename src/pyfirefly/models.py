"""Models for the Firefly API."""

from __future__ import annotations

from dataclasses import dataclass, field

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class About(DataClassORJSONMixin):
    """Model for the Firefly About information."""

    version: str = field(metadata=field_options(alias="version"))
    api_version: str = field(metadata=field_options(alias="api_version"))
    php_version: str = field(metadata=field_options(alias="php_version"))
    os: str = field(metadata=field_options(alias="os"))
    driver: str = field(metadata=field_options(alias="driver"))
