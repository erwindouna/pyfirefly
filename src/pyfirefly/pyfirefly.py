"""Asynchronous Python client for Python Firefly."""

from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self
from urllib.parse import urlparse

from aiohttp import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from pyfirefly.exceptions import (
    FireflyAuthenticationError,
    FireflyConnectionError,
    FireflyError,
    FireflyNotFoundError,
    FireflyTimeoutError,
)
from pyfirefly.models import About

try:
    VERSION = metadata.version(__package__)
except metadata.PackageNotFoundError:
    VERSION = "DEV-0.0.0"


@dataclass
class Firefly:
    """Main class for handling connections with the Python Firefly API."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    def __init__(
        self,
        api_url: str,
        api_key: str,
        *,
        request_timeout: float = 10.0,
        session: ClientSession | None = None,
    ) -> None:
        """Initialize the Firefly object.

        Args:
        ----
            api_url: URL of the Firefly API.
            api_key: API key for authentication.
            request_timeout: Timeout for requests (in seconds).
            session: Optional aiohttp session to use.

        """
        self._api_key = api_key
        self._request_timeout = request_timeout
        self._session = session

        parsed_url = urlparse(api_url)
        self._api_host = parsed_url.hostname or "localhost"
        self._api_scheme = parsed_url.scheme or "http"
        self._api_port = parsed_url.port or 9000

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Python Firefly API.

        Args:
        ----
            uri: Request URI, without '/api/', for example, 'status'.
            method: HTTP method to use.
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the Python Firefly API.

        Raises:
        ------
            Python FireflyAuthenticationError: If the API key is invalid.

        """
        url = URL.build(
            scheme=self._api_scheme,
            host=self._api_host,
            port=self._api_port,
            path="/api/v1/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json, text/plain",
            "User-Agent": f"PythonFirefly/{VERSION}",
            "Authorization": f"Bearer {self._api_key}",
        }

        if self._session is None:
            self._session = ClientSession()
            self._close_session = True

        try:
            async with asyncio.timeout(self._request_timeout):
                response = await self._session.request(
                    method,
                    url,
                    headers=headers,
                    params=params,
                )
                response.raise_for_status()
        except TimeoutError as err:
            msg = f"Timeout error while accessing {method} {url}: {err}"
            raise FireflyTimeoutError(msg) from err
        except ClientResponseError as err:
            if err.status == 401:
                msg = f"Authentication failed for {method} {url}: Invalid API key"
                raise FireflyAuthenticationError(msg) from err
            if err.status == 404:
                msg = f"Resource not found at {method} {url}: {err}"
                raise FireflyNotFoundError(msg) from err
            msg = f"Connection error for {method} {url}: {err}"
            raise FireflyConnectionError(msg) from err
        except (ClientError, socket.gaierror) as err:
            msg = f"Unexpected error during {method} {url}: {err}"
            raise FireflyConnectionError(msg) from err

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Firefly API"
            raise FireflyError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return await response.json()

    async def get_about(self) -> About:
        """Get information about the Firefly server.

        Returns
        -------
            An About object with information about the Firefly server.

        """
        about = await self._request("about")
        return About.from_dict(about["data"])

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Firefly object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
