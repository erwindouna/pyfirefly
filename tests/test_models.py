"""Test the models of the Firefly API client."""

from __future__ import annotations

from typing import TYPE_CHECKING

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from tests import load_fixtures

if TYPE_CHECKING:
    from pyfirefly import Firefly
    from pyfirefly.models import About


async def test_about_model(
    aresponses: ResponsesMockServer,
    firefly_client: Firefly,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the About model."""
    aresponses.add(
        "localhost:9000",
        "/api/v1/about",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("about.json"),
        ),
    )

    about: About = await firefly_client.get_about()
    assert about == snapshot
    await firefly_client.close()
