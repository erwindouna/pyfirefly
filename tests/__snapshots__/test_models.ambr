# serializer version: 1
# name: test_about_model
  About(version='5.8.0-alpha.1', api_version='5.8.0-alpha.1', php_version='8.1.5', os='Linux', driver='mysql')
# ---
# name: test_account_transactions_model
  list([
    Transaction(user=None, transaction_journal_id=None, type='transactions', date=None, order=None, currency_id=None, currency_code=None, currency_symbol=None, currency_name=None, currency_decimal_places=None, foreign_currency_id=None, foreign_currency_code=None, foreign_currency_symbol=None, foreign_currency_decimal_places=None, amount=None, foreign_amount=None, description=None, source_id=None, source_name=None, source_iban=None, source_type=None, destination_id=None, destination_name=None, destination_iban=None, destination_type=None, budget_id=None, budget_name=None, category_id=None, category_name=None, bill_id=None, bill_name=None, reconciled=None, notes=None, tags=None, internal_reference=None, external_id=None, external_url=None, original_source=None, recurrence_id=None, recurrence_total=None, recurrence_count=None, bunq_payment_id=None, import_hash_v2=None, sepa_cc=None, sepa_ct_op=None, sepa_ct_id=None, sepa_db=None, sepa_country=None, sepa_ep=None, sepa_ci=None, sepa_batch_id=None, interest_date=None, book_date=None, process_date=None, due_date=None, payment_date=None, invoice_date=None, latitude=None, longitude=None, zoom_level=None, has_attachments=None),
  ])
# ---
# name: test_account_transactions_model.1
  list([
    Transaction(user=None, transaction_journal_id=None, type='transactions', date=None, order=None, currency_id=None, currency_code=None, currency_symbol=None, currency_name=None, currency_decimal_places=None, foreign_currency_id=None, foreign_currency_code=None, foreign_currency_symbol=None, foreign_currency_decimal_places=None, amount=None, foreign_amount=None, description=None, source_id=None, source_name=None, source_iban=None, source_type=None, destination_id=None, destination_name=None, destination_iban=None, destination_type=None, budget_id=None, budget_name=None, category_id=None, category_name=None, bill_id=None, bill_name=None, reconciled=None, notes=None, tags=None, internal_reference=None, external_id=None, external_url=None, original_source=None, recurrence_id=None, recurrence_total=None, recurrence_count=None, bunq_payment_id=None, import_hash_v2=None, sepa_cc=None, sepa_ct_op=None, sepa_ct_id=None, sepa_db=None, sepa_country=None, sepa_ep=None, sepa_ci=None, sepa_batch_id=None, interest_date=None, book_date=None, process_date=None, due_date=None, payment_date=None, invoice_date=None, latitude=None, longitude=None, zoom_level=None, has_attachments=None),
  ])
# ---
# name: test_accounts_model
  list([
    Account(type='accounts', id='2', attributes=AccountAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', active=False, name='My checking account', type='asset', account_role='defaultAsset', currency_id='12', currency_code='EUR', currency_symbol='$', currency_decimal_places=2, native_currency_id='12', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, current_balance='123.45', native_current_balance='123.45', current_balance_date='2018-09-17T12:46:47+01:00', order=1, notes='Some example notes', monthly_payment_date='2018-09-17T12:46:47+01:00', credit_card_type='monthlyFull', account_number='7009312345678', iban='GB98MIDL07009312345678', bic='BOFAUS3N', virtual_balance='123.45', native_virtual_balance='123.45', opening_balance='-1012.12', native_opening_balance='-1012.12', opening_balance_date='2018-09-17T12:46:47+01:00', liability_type='loan', liability_direction='credit', interest='5.3', interest_period='monthly', current_debt='1012.12', include_net_worth=True, longitude=5.916667, latitude=51.983333, zoom_level=6, last_activity=None)),
    Account(type='accounts', id='3', attributes=AccountAttributes(created_at='2019-01-01T10:00:00+01:00', updated_at='2020-01-01T10:00:00+01:00', active=True, name='Savings Account', type='asset', account_role='savingsAsset', currency_id='13', currency_code='USD', currency_symbol='$', currency_decimal_places=2, native_currency_id='13', native_currency_code='USD', native_currency_symbol='$', native_currency_decimal_places=2, current_balance='5000.00', native_current_balance='5000.00', current_balance_date='2020-01-01T10:00:00+01:00', order=2, notes='Main savings account', monthly_payment_date=None, credit_card_type=None, account_number='1234567890', iban='US12345678901234567890', bic='CITIUS33', virtual_balance='0.00', native_virtual_balance='0.00', opening_balance='1000.00', native_opening_balance='1000.00', opening_balance_date='2019-01-01T10:00:00+01:00', liability_type=None, liability_direction=None, interest='1.2', interest_period='yearly', current_debt=None, include_net_worth=True, longitude=-74.006, latitude=40.7128, zoom_level=8, last_activity=None)),
    Account(type='accounts', id='4', attributes=AccountAttributes(created_at='2021-05-10T09:30:00+01:00', updated_at='2022-05-10T09:30:00+01:00', active=True, name='Credit Card', type='liability', account_role='creditCard', currency_id='14', currency_code='GBP', currency_symbol='£', currency_decimal_places=2, native_currency_id='14', native_currency_code='GBP', native_currency_symbol='£', native_currency_decimal_places=2, current_balance='-250.00', native_current_balance='-250.00', current_balance_date='2022-05-10T09:30:00+01:00', order=3, notes='Credit card account', monthly_payment_date='2022-05-15T09:30:00+01:00', credit_card_type='monthlyFull', account_number='9876543210', iban='GB29NWBK60161331926819', bic='NWBKGB2L', virtual_balance='0.00', native_virtual_balance='0.00', opening_balance='0.00', native_opening_balance='0.00', opening_balance_date='2021-05-10T09:30:00+01:00', liability_type='credit', liability_direction='debit', interest='19.99', interest_period='monthly', current_debt='250.00', include_net_worth=False, longitude=0.1278, latitude=51.5074, zoom_level=10, last_activity=None)),
  ])
# ---
# name: test_bills_model
  list([
    Bill(type='bills', id='2', attributes=BillAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', currency_id='5', currency_code='EUR', currency_symbol='$', currency_decimal_places=2, native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, name='Rent', amount_min='123.45', amount_max='123.45', native_amount_min='123.45', native_amount_max='123.45', date='2018-09-17T12:46:47+01:00', end_date='2018-09-17T12:46:47+01:00', extension_date='2018-09-17T12:46:47+01:00', repeat_freq='monthly', skip=0, active=True, order=1, notes='Some example notes', next_expected_match='2018-09-17T12:46:47+01:00', next_expected_match_diff='today', object_group_id='5', object_group_order=5, object_group_title='Example Group', pay_dates=['2018-09-17T12:46:47+01:00'], paid_dates=[BillPaidDate(transaction_group_id='123', transaction_journal_id='123', date='2018-09-17T12:46:47+01:00')])),
  ])
# ---
# name: test_bills_model.1
  list([
    Bill(type='bills', id='2', attributes=BillAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', currency_id='5', currency_code='EUR', currency_symbol='$', currency_decimal_places=2, native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, name='Rent', amount_min='123.45', amount_max='123.45', native_amount_min='123.45', native_amount_max='123.45', date='2018-09-17T12:46:47+01:00', end_date='2018-09-17T12:46:47+01:00', extension_date='2018-09-17T12:46:47+01:00', repeat_freq='monthly', skip=0, active=True, order=1, notes='Some example notes', next_expected_match='2018-09-17T12:46:47+01:00', next_expected_match_diff='today', object_group_id='5', object_group_order=5, object_group_title='Example Group', pay_dates=['2018-09-17T12:46:47+01:00'], paid_dates=[BillPaidDate(transaction_group_id='123', transaction_journal_id='123', date='2018-09-17T12:46:47+01:00')])),
  ])
# ---
# name: test_budgets_model
  list([
    Budget(type='budgets', id='2', attributes=BudgetAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', name='Bills', active=False, notes='Some notes', order=5, auto_budget_type='reset', currency_id='12', currency_code='EUR', currency_symbol='$', currency_decimal_places=2, native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, auto_budget_amount='-1012.12', native_auto_budget_amount='-1012.12', auto_budget_period='monthly', spent=[BudgetSpent(sum='123.45', currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2)])),
  ])
# ---
# name: test_budgets_model.1
  list([
    Budget(type='budgets', id='2', attributes=BudgetAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', name='Bills', active=False, notes='Some notes', order=5, auto_budget_type='reset', currency_id='12', currency_code='EUR', currency_symbol='$', currency_decimal_places=2, native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, auto_budget_amount='-1012.12', native_auto_budget_amount='-1012.12', auto_budget_period='monthly', spent=[BudgetSpent(sum='123.45', currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2)])),
  ])
# ---
# name: test_category_model
  Category(type='categories', id='2', attributes=CategoryAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', name='Lunch', notes='Some example notes', native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, spent=[CategoryAmount(currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2, sum='-12423.45')], earned=[CategoryAmount(currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2, sum='123.45')]))
# ---
# name: test_category_model.1
  Category(type='categories', id='2', attributes=CategoryAttributes(created_at='2018-09-17T12:46:47+01:00', updated_at='2018-09-17T12:46:47+01:00', name='Lunch', notes='Some example notes', native_currency_id='5', native_currency_code='EUR', native_currency_symbol='$', native_currency_decimal_places=2, spent=[CategoryAmount(currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2, sum='-12423.45')], earned=[CategoryAmount(currency_id='5', currency_code='USD', currency_symbol='$', currency_decimal_places=2, sum='123.45')]))
# ---
